bot/src/modules/level/lib/level.ts

```diff
@@ -0,0 +1,262 @@
+      if (xp >= this.levelArray[index].total) continue;
+      currentLevel = index === 0 ? 0 : index - 1;
+      break;
+    }
+
+    if (!currentLevel && currentLevel !== 0)
+      throw "Couldn't calculate level.";
+
+    const current = this.levelArray[currentLevel];
+
+    return {
+      totalXP: xp,
+      level: current.level,
+      levelXP: current.xp,
+      currentXP: xp - current.total,
+    };
+  }
+
+  public static getXPFromLevel(lvl: number) {
+    return 5 * (lvl * lvl) + 50 * lvl + 100;
+  }
+
+  private static generateLevelArray(maxLevel: number) {
+    const array: LevelArray[] = [];
+    let total = 0;
+
+    for (let i = 0; i < maxLevel + 1; i++) {
+      const xp = LevelController.getXPFromLevel(i);
+      // total is the xp from 0, xp is to the next level.
+      array.push({ total, xp, level: i });
+      total += xp;
+    }
+
+    return array;
+  }
+
+  public static async getLevelData(guildId: string, userId: string) {
+    let levelData = await state.db.level.findUnique({
+      where: {
+        userId_guildId: { guildId, userId },
+      },
+    });
+
+    if (!levelData)
+      levelData = await state.db.level.create({
+        data: {
+          user: connectOrCreate(userId),
+          guild: connectOrCreate(guildId),
+        },
+      });
+
+    return levelData;
+  }
+
+  // user related
+  public isUserTimedOut(member: GuildMember, timeout = 60 * 1000): boolean {
+    const id = `${member.guild.id}-${member.id}`;
+    const lastTimeout = this.lastXPgrant.get(id);
+    if (lastTimeout && Date.now() - timeout < lastTimeout) return false;
+    this.lastXPgrant.set(id, Date.now());
+    return true;
+  }
+
+  public setUserTimeout(member: GuildMember) {
+    const id = `${member.guild.id}-${member.id}`;
+    this.lastXPgrant.set(id, Date.now());
+  }
+
+  public async ChangeUserXP(guildId: string, userId: string, xp = LevelController.getRandomXP()) {
+    const oldData = await LevelController.getLevelData(guildId, userId);
+    const newData = await state.db.level.update({
+      where: { userId_guildId: { guildId: oldData.guildId, userId: oldData.userId } },
+      data: { experience: oldData.experience + xp },
+    });
+
+    const newLevel = this.getLevelFromXP(newData.experience);
+    const oldLevel = this.getLevelFromXP(oldData.experience);
+
+    const levelDelta = newLevel.level - oldLevel.level;
+
+    return { oldData, newData, oldLevel, newLevel, levelDelta };
+  }
+
+  public async vcLoop() {
+    let guilds = state.client.guilds.cache;
+    guilds = guilds.filter((x) => state.guilds.get(x.id)?.levelVCEnabled);
+
+    for (const guild of guilds.values()) {
+      const channels = guild.channels.cache.filter((x) => x.type === ChannelType.GuildVoice) as Collection<string, VoiceChannel>;
+
+      for (const channel of channels.values()) {
+        const members = channel.members;
+        // Filter out bots and deafened users.
+        members.filter((x) => !x.user.bot);
```

Comment:
Why have two different styles of filter here? (One with "(x)", and one with "member"?)